### PR TITLE
feat(across-relayer): Add ability to finalize L2->L1 transfer actions

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

This PR does three things:
- Adds the ability for the relayer to send L2->L1 relay actions over the canonical Optimism/Arbitrum bridge. In this implementation, this tx type will be generated as soon as there are funds to be relayed.
- Refactors Smart contract imports to remove a temp library
- Adds a modification to contract interfaces in the `BridgePool` to include the `bridgeTokens` method to simplify implementers interface.

This may be unnecessary for some PRs. Catch-all for detailed explanations about the implementation decisions and implications of the change.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

Fixes #XXXX

### CODE REVIEW

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (chrismaree) [inline packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts:1]: these tests are somewhat bloated. The reason is they'll need to consider the implementation of the `checkForFinalizedCanonicalRelaysAndFinalize` method, which will require all the L1 contracts deployed here.

File: packages/core/contracts/insured-bridge/BridgeDepositBox.sol
Comment (chrismaree) [inline packages/core/contracts/insured-bridge/BridgeDepositBox.sol:7]: we should have removed this before.

File: packages/core/contracts/insured-bridge/BridgeDepositBox.sol
Comment (chrismaree) [inline packages/core/contracts/insured-bridge/BridgeDepositBox.sol:144]: this is the virtual interface I added. this means that implementers only need to use `BridgeDepositBox` when calling methods and can ignore chain-specific interfaces.

Note that all other important methods in this file contain the same type of interface

File: packages/core/contracts/insured-bridge/BridgeDepositBox.sol
Comment (chrismaree) [inline packages/core/contracts/insured-bridge/BridgeDepositBox.sol:195]: somewhat unfortunate artifact of the transfer helper removal is the need to cast these addresses to interfaces.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:16]: new class. will add comments soon.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:23]: This is cool to think of this bot as part of the `AcrossRelayer` bot but just provides another feature, which is x-domain finalizing!

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:25]: Although its inconsistent to query events in bots in this repo, I think this is a performance optimization as we don't want to query `WhitelistToken` in the InsuredBridgeL1Client

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:27]: We should ideally parameterize the `fromBlock` as `this.firstBlockToSearch` and `toBlock` as `this.endingBlockNumber` so that on each subsequent run, this bot only looks up new events.

Perhaps we should build a `BridgeAdminClient` that handles event processing....

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:31]: nice comment

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:37]: Very nicely written. +1 parallelization

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:41]:
```suggestion
    const bridgeableL2Tokens = whitelistedL2Tokens.filter((, index) => canBridge[index]);
```

I think this works?

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:23]: I think you should emit the L1 client's bridge admin address somewhere, perhaps in this log

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:48]: nit: Just rename to `nonce`. Nonce and counter are the same thing no?

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:55]: Add a comment that the second param is `l2Gas` and is unused for both Arbitrum and Optimism for now

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:34]: Do you think we should add an `if` statement that checks if the bridge deposit box's balance is > some threshold? If its too small then perhaps we hold off bridging to save gas? I know L2 gas is negligible but still

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:52]: nit: put the logic inside this `try{}` block into an internal method `_bridgeL2Token` to make this function more readable and easier to maintain.

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts:216]: Can you add a set of test where there are two L2 whitelisted tokens:
- Both L2 tokens can be bridged
- Only one can be bridged because `canBridge` returns false for one of them so only one bridging action is sent
- Both L2 tokens can be bridged but one fails to send for some reason. For example, you could whitelist one L2 token that always reverts on `safeTransferFrom`? I know this is annoying but definitely increases coverage.

**Review by nicholaspai — COMMENTED** [review]
Amazing work, Chris. I left a lot of comments that but this is going to be super useful.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:25]: correct. it's also way less code. I could move it into a helper perhaps, like that of the index helpers.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:41]: nope. I tried that out but ts gave an error. I think a more correct way would be to use the `_` as the variable place holder? wdyt

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:34]: yup! I will change this to be greater than x% of the amount of liquid + utilised funds on L1. i.e the funds waiting on to be bridged on the canonical bridge are greater than 5% of the funds on L1 then, and only then, bridge

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:23]: This is emitted in the start up log of the the bot in the here: https://github.com/UMAprotocol/protocol/blob/2fc8fe4237156e68a02bf15ff9f981eed0598137/packages/insured-bridge-relayer/src/index.ts#L31

do you want it again here?

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:27]: we _could_....is it worth it though for this one look back? also this is running on L1 and infura has no problem in calling this kind of method.

we could also store the deployment block of the bridge admin in a config file to decrease how far back this has to look?

WDYT

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:48]: done.

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (mrice32) [inline packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts:244]: extra console log

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (mrice32) [inline packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts:251]: console log

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (mrice32) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:27]: Unrelated to the discussion above: why do we need a toBlock at all?

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (mrice32) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:50]: sorry, where are we incrementing the nonce? And I thought web3 would track the nonce internally to ensure that it incremented after each send?

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (mrice32) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:53]: This is an async function. Shouldn't we be awaiting or at least tracking the promises?

**Review by mrice32 — COMMENTED** [review]
Looks good! A few minor comments

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:53]: should be, my mistake.

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts:244]: thanks

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts:251]: got it

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:50]: it should...but it does not work correctly on L2 because the transactions process so quickly 🤖

Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:50]: it is incremented in the `bridgeL2Token` method.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:27]: @mrice32 good point! removing this and leaving it just with the `fromBLock` works just fine.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:22]: oh good point, nevermind

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:56]: It's kind of weird to return the `nonce`. I assume the `nonce` is getting incremented inside `_bridgeL2Token` which is potentially dangerous since it's not made explicit. I think it might be more clear to increment `nonce++` before calling `await this._bridgeL2Token()` and then passing in the new `nonce`.

This would make this logic more functional and easier to reason about

File: packages/insured-bridge-relayer/src/RelayerConfig.ts
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/RelayerConfig.ts:108]: See fix [here](https://github.com/UMAprotocol/protocol/pull/3593/files), this is eventually going to have to change to `CROSS_DOMAIN_FINALIZER_ENABLED === "true" ? true : false`

**Review by nicholaspai — COMMENTED** [review]
Have one comment about making the `_bridgeL2Tokens` loop be more functional programming and easier to reason about

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts]: this has been improved somewhat.

File: packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/CrossDomainFinalizer.ts:56]: post increment nonce

**Review by nicholaspai — APPROVED** [review]
LGTM after you finish matt's open comments

File: packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts
Comment (chrismaree) [reply packages/insured-bridge-relayer/test/CrossDomainFinalizer.ts]: done

**Pull request merged**
